javascript/lib/types.ts

```diff
@@ -0,0 +1,328 @@
+        var length = buf_meta.length().low/this.TypedArray.BYTES_PER_ELEMENT;
+        bufReader.index += 1|0;
+        return new this.TypedArray(buffer, offset, length);
+    }
+
+    getDataView() {
+        return this.dataView;
+    }
+
+    toString() {
+        return this.dataView.toString();
+    }
+
+    slice(start, end) {
+        return this.dataView.slice(start, end);
+    }
+}
+
+class NullableSimpleVector<T extends ArrayView> extends SimpleVector<T> {
+    private validityView: BitArray;
+
+    get(i: number) {
+        if (this.validityView.get(i)) return this.dataView[i];
+        else                          return null
+    }
+
+    loadBuffers(recordBatch, buffer, bufReader, baseOffset) {
+        this.validityView = Vector.loadValidityBuffer(recordBatch, buffer, bufReader, baseOffset);
+        this.dataView = this.loadDataBuffer(recordBatch, buffer, bufReader, baseOffset);
+    }
+
+}
+
+class Uint8Vector   extends SimpleVector<Uint8Array>   { constructor(name: string) { super(Uint8Array,  name);  }; }
+class Uint16Vector  extends SimpleVector<Uint16Array>  { constructor(name: string) { super(Uint16Array, name);  }; }
+class Uint32Vector  extends SimpleVector<Uint32Array>  { constructor(name: string) { super(Uint32Array, name);  }; }
+class Int8Vector    extends SimpleVector<Uint8Array>   { constructor(name: string) { super(Uint8Array,  name);  }; }
+class Int16Vector   extends SimpleVector<Uint16Array>  { constructor(name: string) { super(Uint16Array, name);  }; }
+class Int32Vector   extends SimpleVector<Uint32Array>  { constructor(name: string) { super(Uint32Array, name);  }; }
+class Float32Vector extends SimpleVector<Float32Array> { constructor(name: string) { super(Float32Array, name); }; }
+class Float64Vector extends SimpleVector<Float64Array> { constructor(name: string) { super(Float64Array, name); }; }
+
+class NullableUint8Vector   extends NullableSimpleVector<Uint8Array>   { constructor(name: string) { super(Uint8Array,  name);  }; }
+class NullableUint16Vector  extends NullableSimpleVector<Uint16Array>  { constructor(name: string) { super(Uint16Array, name);  }; }
+class NullableUint32Vector  extends NullableSimpleVector<Uint32Array>  { constructor(name: string) { super(Uint32Array, name);  }; }
+class NullableInt8Vector    extends NullableSimpleVector<Uint8Array>   { constructor(name: string) { super(Uint8Array,  name);  }; }
+class NullableInt16Vector   extends NullableSimpleVector<Uint16Array>  { constructor(name: string) { super(Uint16Array, name);  }; }
+class NullableInt32Vector   extends NullableSimpleVector<Uint32Array>  { constructor(name: string) { super(Uint32Array, name);  }; }
+class NullableFloat32Vector extends NullableSimpleVector<Float32Array> { constructor(name: string) { super(Float32Array, name); }; }
+class NullableFloat64Vector extends NullableSimpleVector<Float64Array> { constructor(name: string) { super(Float64Array, name); }; }
+
```

Comment:
Yeah, let's definitely get rid of these, doesn't need to be done in this patch though. Whether a field is not nullable or has null count zero, we can skip interacting with the validity bitmap. In C++, for non-nullable fields the bitmap is null so any calls to `IsNull` always return false (https://github.com/apache/arrow/blob/master/cpp/src/arrow/array.h#L92). 

We're going to want to implement the visitor pattern for loading and unloading the arrays (or vectors as they're called here and in Java) in TypeScript. see https://github.com/torokmark/design_patterns_in_typescript/blob/master/visitor/visitor.ts for an example of visitor pattern in TS